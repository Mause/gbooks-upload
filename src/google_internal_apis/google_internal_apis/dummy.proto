import "google/protobuf/wrappers.proto";

message DummyMessage {
  optional google.protobuf.StringValue dummy_field = 1;
}
