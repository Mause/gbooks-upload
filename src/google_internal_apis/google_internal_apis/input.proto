import "google/protobuf/timestamp.proto";
import "google/protobuf/wrappers.proto";

message Tag {
  required string name = 1;
  required string tag_id = 2;
  optional google.protobuf.Timestamp created_at = 3;
}

message Tagged {
  required string book_id = 1;
  required string tag_id = 2;
  required google.protobuf.Timestamp tagged_at = 3;
}

message TagsResponse {
  repeated Tag tags = 1;
  repeated Tagged tagged_items = 2;
}
message TagRequest {
  repeated Tagged tagged_items = 1;
}

message LibraryDocumentResponse {
  required LibraryDocument library_documents = 1;
}

message LibraryDocument {
  required string book_id = 1;
  message Book {
    required string title = 1;
    repeated string authors = 2;
    required string publisher = 3;
    required string published_date = 4;
    required string description = 5;
    required int32 page_count = 6;
    required string version = 7;
    repeated CoverImage cover_images = 8;
    required string language = 9;
    required string url = 10;

    // [0,
    // [1, 0],
    // 1,
    // 1,
    // [
    //	'No Dogs in Philly',
    //  [['u4P3GgAAABBNmM', 1, 1, None, 2, 'Special Sin']],
    //  '1'
    // ],
    // [0, 0],
    // None,
    // [['geo', '7']],
    // None,
    // [4.147059, '102'],
    // [[['Fantasy', 'Dark Fantasy']],
    // [['Horror']],
    // [['Science Fiction', 'Alien Contact']],
    // [['Science Fiction', 'Cyberpunk']],
    // [['Visionary & Metaphysical']]],
    // None,
    // None,
    // None,
    // None,
    // None,
    // 1,
    // None,
    // 1,
    // 1]
    required int32 unknown1 = 11;
    repeated int32 unknown2s = 12;
    required int32 unknown3 = 13;
    required int32 unknown4 = 14;
    required Series series = 15;
    repeated int32 unknown5s = 16;
    required google.protobuf. StringValue unknown6 = 17;
    repeated Geo geos = 18;
    required google.protobuf.StringValue unknown7 = 19;
    required Rating rating = 20;
    repeated BookGenre genres = 21;

    message BookGenre {
      required Genre genre = 1;
    }

    message Geo {
      required string geo = 1;
      required string geo_id = 2;
    }

    message Genre {
      required string genre = 1;
      optional string subgenre = 2;
    }

    message Rating {
      required float rating = 1;
      required string number_of_reviews = 2;
    }

    message Series {
      required string this_book_title = 1;
      repeated SingleSeries series=2;
      required string book_in_series = 3;
    }
    message SingleSeries {
      required string series_id = 1;
      required int32 unknown1 = 2;
      required int32 unknown2 = 3;
      required google.protobuf.StringValue unknown3 = 4;
      required int32 unknown4 = 5;
      required string series_name = 6;
    }


    message CoverImage {
      required string url = 1;
      required int32 scale = 2;
      required string color = 3;
      required int32 width = 4;
      required int32 height = 5;
    }
  }
  required Book book = 2;
}
