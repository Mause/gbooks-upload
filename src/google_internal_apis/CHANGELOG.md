# Changelog

## [0.3.0](https://github.com/Mause/gbooks-upload/compare/v0.2.0...v0.3.0) (2025-02-04)


### Features

* add from_datetime helper ([dce940f](https://github.com/Mause/gbooks-upload/commit/dce940f55085c99703970b9fbaede9c85a3ecb95))


### Documentation

* add code sample ([367ca55](https://github.com/Mause/gbooks-upload/commit/367ca55763ed9451c4399e37cdf1b8a70509c8ef))
* add google_internal_apis README.md ([48b0d3c](https://github.com/Mause/gbooks-upload/commit/48b0d3ca9b8477afaffe329dbe9e521625b48c88))
* add some docstrings ([f15664a](https://github.com/Mause/gbooks-upload/commit/f15664ad39336234abb969d5097e85d15f6bba91))

## [0.2.0](https://github.com/Mause/gbooks-upload/compare/0.1.0...v0.2.0) (2025-02-04)


### Features

* add dump function ([fde3ebf](https://github.com/Mause/gbooks-upload/commit/fde3ebfcf7591f1ee78dec238cd649da4345b71c))
* add from_datetime helper ([dce940f](https://github.com/Mause/gbooks-upload/commit/dce940f55085c99703970b9fbaede9c85a3ecb95))


### Bug Fixes

* improve optional support ([02ce14b](https://github.com/Mause/gbooks-upload/commit/02ce14b8dfcc17327e1e96c0b1f5c18e784951cd))


### Documentation

* add some docstrings ([f15664a](https://github.com/Mause/gbooks-upload/commit/f15664ad39336234abb969d5097e85d15f6bba91))

## 0.1.0 (2025-02-04)


### Features

* add dump function ([fde3ebf](https://github.com/Mause/gbooks-upload/commit/fde3ebfcf7591f1ee78dec238cd649da4345b71c))
* add from_datetime helper ([dce940f](https://github.com/Mause/gbooks-upload/commit/dce940f55085c99703970b9fbaede9c85a3ecb95))


### Bug Fixes

* improve optional support ([02ce14b](https://github.com/Mause/gbooks-upload/commit/02ce14b8dfcc17327e1e96c0b1f5c18e784951cd))


### Documentation

* add some docstrings ([f15664a](https://github.com/Mause/gbooks-upload/commit/f15664ad39336234abb969d5097e85d15f6bba91))
