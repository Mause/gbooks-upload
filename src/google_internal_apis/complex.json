[
  [
    "hdYTCAAAQBAJ",
    [
      "No Dogs in Philly",
      [
        "Andy Futuro"
      ],
      "June Day Press",
      "2014-06-08",
      "&quot;This is one of the most enjoyable books I have read in some time.\u00a0 It is in the running for one of the top 5 books for 2015.\u00a0 I know the year is not even half over, but the book is that good.&quot; -Word Refiner <br><br>An OpenBooks science fiction best seller of monsters, Gods, and aliens, unlike anything you&#39;ve ever read.<br><br>Philadelphia. Elzi on every corner, cops just itching to crack a skull, and the Gaespora lordin&#39; it up in their high towers while the rest of the filth dribbled down the sewer. Saru had a way out. All she had to do was find the girl, one skinny stray with blue, blue eyes--bluer than anyone had ever seen--and ten million fat bucks were hers. Except someone was killing blue-eyed girls, and they were A-list, major-league, cold-sweat effective. And something about the end of all existence if she failed. <br>\u00a0 <br>No Dogs in Philly is a Lovecraftian Cyberpunk Noir with aliens, monsters, extra-dimensional death Gods and a hardboiled female protagonist. It tackles questions of existence and the role humans play in this particular universe. <br><br>Rated R for strong language, mentions of sex, and graphic violence. Contains intense horror and potentially disturbing imagery. <br><br>No Dogs in Philly may appeal to fans of H.P. Lovecraft, Neal Stephenson, China Mieville, Dan Simmons, Gyo, Tank Girl, Swamp Thing, Spawn, science fiction, horror, cyberpunk, absurdism, urban fantasy, new weird, weird fiction, slipstream, and speculative fiction. <br><br>Keywords: dark books, gritty books, noir books, horror books, weird books, strange books, unique books, different books, controversial books, challenging books, surprising books, unsettling books, disturbing books, bizarre books, unusual books, scary books, absurd books, crazy books, violent books, bloody books, gory books, books like sandman slim, books like snowcrash, books like neal stephenson, books like neil stephenson, books like lovecraft, books like william gibson, books like gibson, books with dark covers, books with female heroines, black humor, quick reads, weird book series, books that make you think, gritty thrillers, dark thrillers, sci fi thrillers, weird thrillers, alien books, monster books, metaphysical books.<br>",
      168,
      "2.7.6.0.full.2",
      [
        [
          "https://books.google.com/books/publisher/content/images/frontcover/hdYTCAAAQBAJ",
          1,
          "#0458c6",
          1600,
          2560
        ]
      ],
      "en",
      "https://play.google.com/store/books/details?id=hdYTCAAAQBAJ",
      0,
      [
        1,
        0
      ],
      1,
      1,
      [
        "No Dogs in Philly",
        [
          [
            "u4P3GgAAABBNmM",
            1,
            1,
            null,
            2,
            "Special Sin"
          ]
        ],
        "1"
      ],
      [
        0,
        0
      ],
      null,
      [
        [
          "geo",
          "7"
        ]
      ],
      null,
      [
        4.147059,
        "102"
      ],
      [
        [
          [
            "Fantasy",
            "Dark Fantasy"
          ]
        ],
        [
          [
            "Horror"
          ]
        ],
        [
          [
            "Science Fiction",
            "Alien Contact"
          ]
        ],
        [
          [
            "Science Fiction",
            "Cyberpunk"
          ]
        ],
        [
          [
            "Visionary & Metaphysical"
          ]
        ]
      ],
      null,
      null,
      null,
      null,
      null,
      1,
      null,
      1,
      1
    ],
    [
      null,
      "1484205137000000",
      null,
      null,
      null,
      null,
      3,
      "1480857699000000",
      null,
      null,
      null,
      null,
      [
        0
      ],
      1,
      null,
      1,
      null,
      null,
      [
        [
          "GBS.PT137.w.6.0.4",
          "1484205137000000"
        ],
        [
          1,
          49128,
          49128,
          "0"
        ]
      ],
      null,
      null,
      null,
      null,
      1,
      null,
      0,
      1
    ],
    [
      null,
      1,
      [
        "AUD",
        "0"
      ],
      "https://play.google.com/store/books/details?id=hdYTCAAAQBAJ&rdid=book-hdYTCAAAQBAJ&rdot=1",
      [
        [
          null,
          [
            "AUD",
            "0"
          ],
          [
            "AUD",
            "0"
          ],
          1,
          0,
          0
        ],
        [
          null,
          [
            "AUD",
            "0"
          ],
          [
            "AUD",
            "0"
          ],
          1,
          0,
          1
        ]
      ]
    ],
    [
      3,
      null,
      2,
      null,
      [
        null,
        0
      ],
      null,
      1,
      null,
      null,
      [
        [
          [
            1,
            "https://books.google.com/books/download/No_Dogs_in_Philly.epub?id=hdYTCAAAQBAJ&output=epub&source=gbs_api"
          ]
        ]
      ]
    ]
  ]
]
