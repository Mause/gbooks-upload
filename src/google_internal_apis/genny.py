import logging
import re
from pathlib import Path
from subprocess import check_call
from textwrap import dedent
from typing import NamedTuple

import yaml
from jinja2 import Environment
from rich.logging import RichHandler

logging.basicConfig(level=logging.INFO, handlers=[RichHandler()])

HERE = Path(__file__).parent


class Method(NamedTuple):
    hostname: str
    service: str
    method: str
    body: str


def get_methods():
    with open(HERE / "endpoints.yaml") as f:
        return yaml.load(f, yaml.SafeLoader)


def key(x):
    return x.hostname, x.service


template = """
class {{classname}}Rpc(RpcService):
    service = "{{service}}"

    def __init__(self, creds, client):
        super().__init__(creds, client)
        self.hostname = "{{hostname}}"

    {% for method in methods %}
    async def {{method.method | lower}}(self, data={% if method.body %}
        {{method.body | repr}}
        {% else %}
        None
        {% endif %}):
        return await self._call_rpc({{method.method | repr}}, data=data)
    {% endfor %}


"""


def lower(s):
    return re.sub("([a-z])([A-Z])", r"\1_\2", s).lower()


env = Environment()
env.filters["lower"] = lower
env.filters["repr"] = repr
t = env.from_string(template)


def main():
    methods = get_methods()

    OUT = HERE / "google_internal_apis/endpoints.py"

    with open(OUT, "w") as f:
        f.write(
            dedent("""
        from .ghunter import RpcService

        """)
        )

        for hostname, services in methods.items():
            for service, methods in services.items():
                if isinstance(methods, str):
                    logging.warning(f"Skipping {hostname}.{service}")
                    continue
                f.write(
                    t.render(
                        classname=service.split(".")[-1],
                        hostname=hostname,
                        service=service,
                        methods=methods["methods"],
                    )
                )
    check_call(["ruff", "format", OUT])
    check_call(["ruff", "check", OUT, "--fix"])


if __name__ == "__main__":
    main()
