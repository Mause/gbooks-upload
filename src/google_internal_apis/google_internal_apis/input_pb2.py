# -*- coding: utf-8 -*-
# Generated by the protocol buffer compiler.  DO NOT EDIT!
# source: google_internal_apis/input.proto
"""Generated protocol buffer code."""
from google.protobuf.internal import builder as _builder
from google.protobuf import descriptor as _descriptor
from google.protobuf import descriptor_pool as _descriptor_pool
from google.protobuf import symbol_database as _symbol_database
# @@protoc_insertion_point(imports)

_sym_db = _symbol_database.Default()


from google.protobuf import timestamp_pb2 as google_dot_protobuf_dot_timestamp__pb2
from google.protobuf import wrappers_pb2 as google_dot_protobuf_dot_wrappers__pb2


DESCRIPTOR = _descriptor_pool.Default().AddSerializedFile(b'\n google_internal_apis/input.proto\x1a\x1fgoogle/protobuf/timestamp.proto\x1a\x1egoogle/protobuf/wrappers.proto\"S\n\x03Tag\x12\x0c\n\x04name\x18\x01 \x02(\t\x12\x0e\n\x06tag_id\x18\x02 \x02(\t\x12.\n\ncreated_at\x18\x03 \x01(\x0b\x32\x1a.google.protobuf.Timestamp\"X\n\x06Tagged\x12\x0f\n\x07\x62ook_id\x18\x01 \x02(\t\x12\x0e\n\x06tag_id\x18\x02 \x02(\t\x12-\n\ttagged_at\x18\x03 \x02(\x0b\x32\x1a.google.protobuf.Timestamp\"A\n\x0cTagsResponse\x12\x12\n\x04tags\x18\x01 \x03(\x0b\x32\x04.Tag\x12\x1d\n\x0ctagged_items\x18\x02 \x03(\x0b\x32\x07.Tagged\"+\n\nTagRequest\x12\x1d\n\x0ctagged_items\x18\x01 \x03(\x0b\x32\x07.Tagged\"F\n\x17LibraryDocumentResponse\x12+\n\x11library_documents\x18\x01 \x02(\x0b\x32\x10.LibraryDocument\"\xc0\t\n\x0fLibraryDocument\x12\x0f\n\x07\x62ook_id\x18\x01 \x02(\t\x12#\n\x04\x62ook\x18\x02 \x02(\x0b\x32\x15.LibraryDocument.Book\x1a\xf6\x08\n\x04\x42ook\x12\r\n\x05title\x18\x01 \x02(\t\x12\x0f\n\x07\x61uthors\x18\x02 \x03(\t\x12\x11\n\tpublisher\x18\x03 \x02(\t\x12\x16\n\x0epublished_date\x18\x04 \x02(\t\x12\x13\n\x0b\x64\x65scription\x18\x05 \x02(\t\x12\x12\n\npage_count\x18\x06 \x02(\x05\x12\x0f\n\x07version\x18\x07 \x02(\t\x12\x36\n\x0c\x63over_images\x18\x08 \x03(\x0b\x32 .LibraryDocument.Book.CoverImage\x12\x10\n\x08language\x18\t \x02(\t\x12\x0b\n\x03url\x18\n \x02(\t\x12\x10\n\x08unknown1\x18\x0b \x02(\x05\x12\x11\n\tunknown2s\x18\x0c \x03(\x05\x12\x10\n\x08unknown3\x18\r \x02(\x05\x12\x10\n\x08unknown4\x18\x0e \x02(\x05\x12,\n\x06series\x18\x0f \x02(\x0b\x32\x1c.LibraryDocument.Book.Series\x12\x11\n\tunknown5s\x18\x10 \x03(\x05\x12.\n\x08unknown6\x18\x11 \x02(\x0b\x32\x1c.google.protobuf.StringValue\x12\'\n\x04geos\x18\x12 \x03(\x0b\x32\x19.LibraryDocument.Book.Geo\x12.\n\x08unknown7\x18\x13 \x02(\x0b\x32\x1c.google.protobuf.StringValue\x12,\n\x06rating\x18\x14 \x02(\x0b\x32\x1c.LibraryDocument.Book.Rating\x12/\n\x06genres\x18\x15 \x03(\x0b\x32\x1f.LibraryDocument.Book.BookGenre\x1a\x37\n\tBookGenre\x12*\n\x05genre\x18\x01 \x02(\x0b\x32\x1b.LibraryDocument.Book.Genre\x1a\"\n\x03Geo\x12\x0b\n\x03geo\x18\x01 \x02(\t\x12\x0e\n\x06geo_id\x18\x02 \x02(\t\x1a(\n\x05Genre\x12\r\n\x05genre\x18\x01 \x02(\t\x12\x10\n\x08subgenre\x18\x02 \x01(\t\x1a\x33\n\x06Rating\x12\x0e\n\x06rating\x18\x01 \x02(\x02\x12\x19\n\x11number_of_reviews\x18\x02 \x02(\t\x1am\n\x06Series\x12\x17\n\x0fthis_book_title\x18\x01 \x02(\t\x12\x32\n\x06series\x18\x02 \x03(\x0b\x32\".LibraryDocument.Book.SingleSeries\x12\x16\n\x0e\x62ook_in_series\x18\x03 \x02(\t\x1a\x9c\x01\n\x0cSingleSeries\x12\x11\n\tseries_id\x18\x01 \x02(\t\x12\x10\n\x08unknown1\x18\x02 \x02(\x05\x12\x10\n\x08unknown2\x18\x03 \x02(\x05\x12.\n\x08unknown3\x18\x04 \x02(\x0b\x32\x1c.google.protobuf.StringValue\x12\x10\n\x08unknown4\x18\x05 \x02(\x05\x12\x13\n\x0bseries_name\x18\x06 \x02(\t\x1aV\n\nCoverImage\x12\x0b\n\x03url\x18\x01 \x02(\t\x12\r\n\x05scale\x18\x02 \x02(\x05\x12\r\n\x05\x63olor\x18\x03 \x02(\t\x12\r\n\x05width\x18\x04 \x02(\x05\x12\x0e\n\x06height\x18\x05 \x02(\x05')

_globals = globals()
_builder.BuildMessageAndEnumDescriptors(DESCRIPTOR, _globals)
_builder.BuildTopDescriptorsAndMessages(DESCRIPTOR, 'google_internal_apis.input_pb2', _globals)
if _descriptor._USE_C_DESCRIPTORS == False:

  DESCRIPTOR._options = None
  _globals['_TAG']._serialized_start=101
  _globals['_TAG']._serialized_end=184
  _globals['_TAGGED']._serialized_start=186
  _globals['_TAGGED']._serialized_end=274
  _globals['_TAGSRESPONSE']._serialized_start=276
  _globals['_TAGSRESPONSE']._serialized_end=341
  _globals['_TAGREQUEST']._serialized_start=343
  _globals['_TAGREQUEST']._serialized_end=386
  _globals['_LIBRARYDOCUMENTRESPONSE']._serialized_start=388
  _globals['_LIBRARYDOCUMENTRESPONSE']._serialized_end=458
  _globals['_LIBRARYDOCUMENT']._serialized_start=461
  _globals['_LIBRARYDOCUMENT']._serialized_end=1677
  _globals['_LIBRARYDOCUMENT_BOOK']._serialized_start=535
  _globals['_LIBRARYDOCUMENT_BOOK']._serialized_end=1677
  _globals['_LIBRARYDOCUMENT_BOOK_BOOKGENRE']._serialized_start=1133
  _globals['_LIBRARYDOCUMENT_BOOK_BOOKGENRE']._serialized_end=1188
  _globals['_LIBRARYDOCUMENT_BOOK_GEO']._serialized_start=1190
  _globals['_LIBRARYDOCUMENT_BOOK_GEO']._serialized_end=1224
  _globals['_LIBRARYDOCUMENT_BOOK_GENRE']._serialized_start=1226
  _globals['_LIBRARYDOCUMENT_BOOK_GENRE']._serialized_end=1266
  _globals['_LIBRARYDOCUMENT_BOOK_RATING']._serialized_start=1268
  _globals['_LIBRARYDOCUMENT_BOOK_RATING']._serialized_end=1319
  _globals['_LIBRARYDOCUMENT_BOOK_SERIES']._serialized_start=1321
  _globals['_LIBRARYDOCUMENT_BOOK_SERIES']._serialized_end=1430
  _globals['_LIBRARYDOCUMENT_BOOK_SINGLESERIES']._serialized_start=1433
  _globals['_LIBRARYDOCUMENT_BOOK_SINGLESERIES']._serialized_end=1589
  _globals['_LIBRARYDOCUMENT_BOOK_COVERIMAGE']._serialized_start=1591
  _globals['_LIBRARYDOCUMENT_BOOK_COVERIMAGE']._serialized_end=1677
# @@protoc_insertion_point(module_scope)
