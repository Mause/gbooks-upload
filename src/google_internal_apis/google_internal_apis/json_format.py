import warnings
from pprint import pformat

from google.protobuf.timestamp_pb2 import Timestamp
from google.protobuf.wrappers_pb2 import StringValue


def dump(message):
    if isinstance(message, Timestamp):
        return str(message.ToMilliseconds())
    if isinstance(message, StringValue):
        return message.value
    fields = message.DESCRIPTOR.fields
    arrays = []
    for field in fields:
        if field.label == field.LABEL_REPEATED:
            if field.type == field.TYPE_MESSAGE:
                arrays.append([dump(item) for item in getattr(message, field.name)])
            else:
                arrays.append(getattr(message, field.name))
        elif field.label == field.LABEL_REQUIRED:
            if field.type == field.TYPE_MESSAGE:
                arrays.append(dump(getattr(message, field.name)))
            else:
                arrays.append(getattr(message, field.name))
        elif field.label == field.LABEL_OPTIONAL:
            if field.type == field.TYPE_MESSAGE:
                arrays.append(dump(getattr(message, field.name)))
            else:
                arrays.append(getattr(message, field.name))
        else:
            raise ValueError("Unknown label")
    return arrays


def repeated(message, field, value):
    if field.type == field.TYPE_MESSAGE:
        for array in value:
            parse(array, getattr(message, field.name).add())
    else:
        getattr(message, field.name).extend(value)


def required(message, field, value):
    if field.type == field.TYPE_MESSAGE:
        parse(value, getattr(message, field.name))
    else:
        setattr(message, field.name, value)


def optional(message, field, value):
    if field.type == field.TYPE_MESSAGE:
        parse(value, getattr(message, field.name))
    else:
        setattr(message, field.name, value)


def parse(arrays, message):
    fields = message.DESCRIPTOR.fields
    if isinstance(message, Timestamp):
        message.FromMilliseconds(int(arrays))
        return message
    if isinstance(message, StringValue):
        if arrays:
            message.MergeFromString(arrays)
        return message

    for field in fields:
        match field.label:
            case field.LABEL_REPEATED:
                value = arrays[field.number - 1]
                repeated(message, field, value)
            case field.LABEL_REQUIRED:
                value = arrays[field.number - 1]
                required(message, field, value)
            case field.LABEL_OPTIONAL:
                if len(arrays) >= field.number:
                    value = arrays[field.number - 1]
                    optional(message, field, value)
            case _:
                raise ValueError("Unknown label")
    remaining = arrays[fields[-1].number :]
    if remaining:
        warnings.warn("Extra fields: " + pformat(remaining))
    return message
