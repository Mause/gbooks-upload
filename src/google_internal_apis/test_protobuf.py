import json
from pathlib import Path

from google.protobuf.json_format import MessageToDict

from google_internal_apis.input_pb2 import LibraryDocumentResponse, TagsResponse
from google_internal_apis.json_format import dump, parse

HERE = Path(__file__).parent


def test_protoc(snapshot):
    original = [
        [
            ["tag_name_1", "tag_id_1", "1"],
            ["tag_name_2", "tag_id_2", "2"],
        ],
        [
            ["book_id_1", "tag_id_1", "1"],
            ["book_id_2", "tag_id_2", "2"],
        ],
    ]
    out = parse(
        original,
        TagsResponse(),
    )
    assert str(out).strip() == snapshot
    assert MessageToDict(out) == snapshot

    assert dump(out) == original


def test_parse_complex(snapshot):
    with open(HERE / "complex.json") as f:
        payload = json.load(f)
    out = parse(payload, LibraryDocumentResponse())
    assert str(out).strip() == snapshot
    assert MessageToDict(out) == snapshot

    # TODO: roundtrip test
