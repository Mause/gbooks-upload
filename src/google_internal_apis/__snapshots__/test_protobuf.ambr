# serializer version: 1
# name: test_parse_complex
  '''
  library_documents {
    book_id: "hdYTCAAAQBAJ"
    book {
      title: "No Dogs in Philly"
      authors: "Andy Futuro"
      publisher: "June Day Press"
      published_date {
        year: 2014
        month: 6
        day: 8
      }
      description: "&quot;This is one of the most enjoyable books I have read in some time.  It is in the running for one of the top 5 books for 2015.  I know the year is not even half over, but the book is that good.&quot; -Word Refiner <br><br>An OpenBooks science fiction best seller of monsters, Gods, and aliens, unlike anything you&#39;ve ever read.<br><br>Philadelphia. Elzi on every corner, cops just itching to crack a skull, and the Gaespora lordin&#39; it up in their high towers while the rest of the filth dribbled down the sewer. Saru had a way out. All she had to do was find the girl, one skinny stray with blue, blue eyes--bluer than anyone had ever seen--and ten million fat bucks were hers. Except someone was killing blue-eyed girls, and they were A-list, major-league, cold-sweat effective. And something about the end of all existence if she failed. <br>  <br>No Dogs in Philly is a Lovecraftian Cyberpunk Noir with aliens, monsters, extra-dimensional death Gods and a hardboiled female protagonist. It tackles questions of existence and the role humans play in this particular universe. <br><br>Rated R for strong language, mentions of sex, and graphic violence. Contains intense horror and potentially disturbing imagery. <br><br>No Dogs in Philly may appeal to fans of H.P. Lovecraft, Neal Stephenson, China Mieville, Dan Simmons, Gyo, Tank Girl, Swamp Thing, Spawn, science fiction, horror, cyberpunk, absurdism, urban fantasy, new weird, weird fiction, slipstream, and speculative fiction. <br><br>Keywords: dark books, gritty books, noir books, horror books, weird books, strange books, unique books, different books, controversial books, challenging books, surprising books, unsettling books, disturbing books, bizarre books, unusual books, scary books, absurd books, crazy books, violent books, bloody books, gory books, books like sandman slim, books like snowcrash, books like neal stephenson, books like neil stephenson, books like lovecraft, books like william gibson, books like gibson, books with dark covers, books with female heroines, black humor, quick reads, weird book series, books that make you think, gritty thrillers, dark thrillers, sci fi thrillers, weird thrillers, alien books, monster books, metaphysical books.<br>"
      page_count: 168
      version: "2.7.6.0.full.2"
      cover_images {
        url: "https://books.google.com/books/publisher/content/images/frontcover/hdYTCAAAQBAJ"
        scale: 1
        color: "#0458c6"
        width: 1600
        height: 2560
      }
      language: "en"
      url: "https://play.google.com/store/books/details?id=hdYTCAAAQBAJ"
      unknown1: 0
      unknown2s: 1
      unknown2s: 0
      unknown3: 1
      unknown4: 1
      series {
        this_book_title: "No Dogs in Philly"
        series {
          series_id: "u4P3GgAAABBNmM"
          unknown1: 1
          unknown2: 1
          unknown4: 2
          series_name: "Special Sin"
        }
        book_in_series: "1"
      }
      unknown5s: 0
      unknown5s: 0
      geos {
        geo: "geo"
        geo_id: "7"
      }
      rating {
        rating: 4.147059
        number_of_reviews: "102"
      }
      genres {
        genre {
          genre: "Fantasy"
          subgenre: "Dark Fantasy"
        }
      }
      genres {
        genre {
          genre: "Horror"
        }
      }
      genres {
        genre {
          genre: "Science Fiction"
          subgenre: "Alien Contact"
        }
      }
      genres {
        genre {
          genre: "Science Fiction"
          subgenre: "Cyberpunk"
        }
      }
      genres {
        genre {
          genre: "Visionary & Metaphysical"
        }
      }
    }
  }
  '''
# ---
# name: test_parse_complex.1
  dict({
    'libraryDocuments': dict({
      'book': dict({
        'authors': list([
          'Andy Futuro',
        ]),
        'coverImages': list([
          dict({
            'color': '#0458c6',
            'height': 2560,
            'scale': 1,
            'url': 'https://books.google.com/books/publisher/content/images/frontcover/hdYTCAAAQBAJ',
            'width': 1600,
          }),
        ]),
        'description': '&quot;This is one of the most enjoyable books I have read in some time.\xa0 It is in the running for one of the top 5 books for 2015.\xa0 I know the year is not even half over, but the book is that good.&quot; -Word Refiner <br><br>An OpenBooks science fiction best seller of monsters, Gods, and aliens, unlike anything you&#39;ve ever read.<br><br>Philadelphia. Elzi on every corner, cops just itching to crack a skull, and the Gaespora lordin&#39; it up in their high towers while the rest of the filth dribbled down the sewer. Saru had a way out. All she had to do was find the girl, one skinny stray with blue, blue eyes--bluer than anyone had ever seen--and ten million fat bucks were hers. Except someone was killing blue-eyed girls, and they were A-list, major-league, cold-sweat effective. And something about the end of all existence if she failed. <br>\xa0 <br>No Dogs in Philly is a Lovecraftian Cyberpunk Noir with aliens, monsters, extra-dimensional death Gods and a hardboiled female protagonist. It tackles questions of existence and the role humans play in this particular universe. <br><br>Rated R for strong language, mentions of sex, and graphic violence. Contains intense horror and potentially disturbing imagery. <br><br>No Dogs in Philly may appeal to fans of H.P. Lovecraft, Neal Stephenson, China Mieville, Dan Simmons, Gyo, Tank Girl, Swamp Thing, Spawn, science fiction, horror, cyberpunk, absurdism, urban fantasy, new weird, weird fiction, slipstream, and speculative fiction. <br><br>Keywords: dark books, gritty books, noir books, horror books, weird books, strange books, unique books, different books, controversial books, challenging books, surprising books, unsettling books, disturbing books, bizarre books, unusual books, scary books, absurd books, crazy books, violent books, bloody books, gory books, books like sandman slim, books like snowcrash, books like neal stephenson, books like neil stephenson, books like lovecraft, books like william gibson, books like gibson, books with dark covers, books with female heroines, black humor, quick reads, weird book series, books that make you think, gritty thrillers, dark thrillers, sci fi thrillers, weird thrillers, alien books, monster books, metaphysical books.<br>',
        'genres': list([
          dict({
            'genre': dict({
              'genre': 'Fantasy',
              'subgenre': 'Dark Fantasy',
            }),
          }),
          dict({
            'genre': dict({
              'genre': 'Horror',
            }),
          }),
          dict({
            'genre': dict({
              'genre': 'Science Fiction',
              'subgenre': 'Alien Contact',
            }),
          }),
          dict({
            'genre': dict({
              'genre': 'Science Fiction',
              'subgenre': 'Cyberpunk',
            }),
          }),
          dict({
            'genre': dict({
              'genre': 'Visionary & Metaphysical',
            }),
          }),
        ]),
        'geos': list([
          dict({
            'geo': 'geo',
            'geoId': '7',
          }),
        ]),
        'language': 'en',
        'pageCount': 168,
        'publishedDate': dict({
          'day': 8,
          'month': 6,
          'year': 2014,
        }),
        'publisher': 'June Day Press',
        'rating': dict({
          'numberOfReviews': '102',
          'rating': 4.147059,
        }),
        'series': dict({
          'bookInSeries': '1',
          'series': list([
            dict({
              'seriesId': 'u4P3GgAAABBNmM',
              'seriesName': 'Special Sin',
              'unknown1': 1,
              'unknown2': 1,
              'unknown4': 2,
            }),
          ]),
          'thisBookTitle': 'No Dogs in Philly',
        }),
        'title': 'No Dogs in Philly',
        'unknown1': 0,
        'unknown2s': list([
          1,
          0,
        ]),
        'unknown3': 1,
        'unknown4': 1,
        'unknown5s': list([
          0,
          0,
        ]),
        'url': 'https://play.google.com/store/books/details?id=hdYTCAAAQBAJ',
        'version': '2.7.6.0.full.2',
      }),
      'bookId': 'hdYTCAAAQBAJ',
    }),
  })
# ---
# name: test_protoc
  '''
  tags {
    name: "tag_name_1"
    tag_id: "tag_id_1"
    created_at {
      nanos: 1000000
    }
  }
  tags {
    name: "tag_name_2"
    tag_id: "tag_id_2"
    created_at {
      nanos: 2000000
    }
  }
  tagged_items {
    book_id: "book_id_1"
    tag_id: "tag_id_1"
    tagged_at {
      nanos: 1000000
    }
  }
  tagged_items {
    book_id: "book_id_2"
    tag_id: "tag_id_2"
    tagged_at {
      nanos: 2000000
    }
  }
  '''
# ---
# name: test_protoc.1
  dict({
    'taggedItems': list([
      dict({
        'bookId': 'book_id_1',
        'tagId': 'tag_id_1',
        'taggedAt': '1970-01-01T00:00:00.001Z',
      }),
      dict({
        'bookId': 'book_id_2',
        'tagId': 'tag_id_2',
        'taggedAt': '1970-01-01T00:00:00.002Z',
      }),
    ]),
    'tags': list([
      dict({
        'createdAt': '1970-01-01T00:00:00.001Z',
        'name': 'tag_name_1',
        'tagId': 'tag_id_1',
      }),
      dict({
        'createdAt': '1970-01-01T00:00:00.002Z',
        'name': 'tag_name_2',
        'tagId': 'tag_id_2',
      }),
    ]),
  })
# ---
